Ffmpeg arrow

vovythevov opened the pull request:
Available features: 
 - Opening/Closing a video
 - Extracting frame information

@mleotta

mleotta commented on arrows/ffmpeg/tests/decode_barcode.h:
```diff
@@ -0,0 +1,76 @@
+static int bc_area = bc_width*bc_height;
+
+// Decode barcodes from test frame images
+uint32_t decode_barcode(kwiver::vital::image_container_sptr img_ct)
```
We should probably share a single copy of this code, and possibly the test data, between this and the VXL tests.  Note the VXL barcode reader is in the master branch now, you may need to merge in master to see it.  We could put the common code in core/tests for now, but really we should create a better place for common tests, data, and utility functions.

mleotta commented on arrows/ffmpeg/ffmpeg_video_input.h:
```diff
@@ -0,0 +1,92 @@
+namespace arrows {
+namespace ffmpeg {
+
+/// Video input using VXL vidl ffmpeg services.
```
update comments to remove VXL references

mleotta commented on arrows/ffmpeg/ffmpeg_video_input.h:
```diff
@@ -0,0 +1,92 @@
+ * \brief \todo
+ */
+
+#ifndef KWIVER_ARROWS_FFMPEG_ffmpeg_video_input_H
```
use all upper case in include guard

mleotta commented on arrows/ffmpeg/ffmpeg_video_input.cxx:
```diff
@@ -0,0 +1,772 @@
+#include <sstream>
+
+extern "C" {
+#if FFMPEG_IN_SEVERAL_DIRECTORIES
```
We don't need many old versions of FFMPEG like VXL does.  The main focus is the version in Fletch.  I'm not sure how long ago this include path change happened, but if it was several years ago it's likely we don't need to support the old way.

linus-sherrill commented on arrows/ffmpeg/ffmpeg_init.h:
```diff
@@ -0,0 +1,43 @@
+#define ffmpeg_init_h_
+
+//: Initialize the ffmpeg codecs.
+//
```
This doesn't look like doxygen format documentation.

linus-sherrill commented on arrows/ffmpeg/ffmpeg_init.cxx:
```diff
@@ -0,0 +1,55 @@
+{
+  static bool initialized = false;
+  if ( ! initialized ) {
+    av_register_all();
```
I'm not sure how this will be used, but are there any race conditions if two threads try to initialize at the same time?

linus-sherrill commented on arrows/ffmpeg/ffmpeg_video_input.h:
```diff
@@ -0,0 +1,92 @@
+ */
+class KWIVER_ALGO_FFMPEG_EXPORT ffmpeg_video_input
+  : public vital::algorithm_impl < ffmpeg_video_input, vital::algo::video_input >
+{
```
Since this is derived from video_input it should support algorithm capabilities.

dstoup commented on CMake/FindFFMPEG.cmake:
```diff
@@ -0,0 +1,89 @@
+
+set( FFMPEG_FOUND "NO" )
+
+find_path( FFMPEG_INCLUDE1_DIR ffmpeg/avcodec.h
```
Since this is kwiver code, should we add some priming of paths from fletch_DIR? We probably want to give preference to that package to ensure they continue to work together.

dstoup commented on CMake/kwiver-depends-ffmpeg.cmake:
```diff
@@ -0,0 +1,10 @@
+  )
+
+if( KWIVER_ENABLE_FFMPEG )
+  find_package( FFMPEG REQUIRED )
```
Do we care about minimum version of FFmpeg at this point?

mleotta replied:
This change could be done in a follow-up branch.

mleotta commented on arrows/ffmpeg/ffmpeg_video_input.cxx:
```diff
@@ -0,0 +1,834 @@
+  *
+  * @return \b A kwiver::vital::image_pixel_traits
+  */
+  pixel_format pixel_format_from_ffmpeg(AVPixelFormat ffmpeg_format) const
```
I suggest simplifying this quite a bit and removing this function and the `pixel_format` enum.  The VXL vidl library was designed to be able to store these various video frame encodings in raw form.  Most of these are packed formats with different resolutions in the different color channels that `vital_image` cannot support.  In our case we should always ask FFMPEG to convert the image into RGB24 for us.  The only exceptions would be
- `RGBA` - create a 4 channel image 
- `GRAY8` - create a 1 channel image
- `MONOWHITE` and `MONOBLACK` - create a 1 channel boolean image

Note that it is quite rare to find videos encoded with these.  Most videos use some variant of YUV.

mleotta commented on arrows/ffmpeg/ffmpeg_video_input.cxx:
```diff
@@ -0,0 +1,834 @@
+  *
+  * @return \b Current frame number.
+  */
+  unsigned int frame_number() const
```
same comment as above.  We should keep a frame counter (how many times you advanced) and return that.  This gets tricky when dealing with seeking.  It may require a novel solution to make seeking reliably frame accurate.

mleotta replied:
Again, I would remove this function.  It has been shown to be broken in VXL and we don't use it in `vidl_ffmpeg_video_input`.  We remember the frame number in a member variable instead.

mleotta commented on arrows/ffmpeg/ffmpeg_video_input.cxx:
```diff
@@ -0,0 +1,834 @@
+    if (format == ffmpeg_video_input::priv::PIXEL_FORMAT_UNKNOWN)
+    {
+      int size = width * height * 3;
+      if (!d->current_image_memory || size != d->current_image_memory->size())
```
We probably don't want to reuse the same image memory for decoding each frame.  This is great for efficiency, but will corrupt any image that an outside user is still holding on to.  At this point we should just allocate a new image for each new frame.  You can probably just make the image directly and let it allocate it's own memory and access the memory instead of needing to allocate the memory and building an image object around it.

If this memory allocation becomes a bottleneck, I recommend creating a special image memory allocator that builds a pool of image image memory objects such that the memory gets returned to the pool instead of deallocated when the last shared point to an image goes out of scope.  However, this optimization is complex and should be put off until it is needed.

mleotta commented on arrows/ffmpeg/ffmpeg_video_input.cxx:
```diff
@@ -0,0 +1,834 @@
+          d->current_image_memory = vital::image_memory_sptr(new vital::image_memory(size));
+        }
+        avpicture_fill(&test_frame, (uint8_t*)d->current_image_memory->data(), enc->pix_fmt, width, height);
+        av_picture_copy(&test_frame, (AVPicture*)d->f_frame, enc->pix_fmt, width, height);
```
This case is okay, but only in those cases where the data is in a recognized format and only needs too be copied: RGB24, RGBA, GRAY8, MONOWHITE, MONOBLACK

mleotta commented on arrows/ffmpeg/ffmpeg_video_input.cxx:
```diff
@@ -0,0 +1,834 @@
+      // Copy the image into contiguous memory.
+      else
+      {
+        if (!d->current_image_memory)
```
I would just store `d->current_image` and not `d->current_image_memory`, and always create a new image for each new frame.  The image memory is kind of redundant to store since it's inside the image.

vovythevov commented on arrows/ffmpeg/ffmpeg_video_input.cxx:
```diff
@@ -0,0 +1,834 @@
+  * @brief Returns the double value to convert from a stream time base to
+  *  a frame number
+  */
+  double stream_time_base_to_frame() const
```
I think we should improve this in a separate branch that improves the seeking in itself.

vovythevov commented:
@chetnieter: If you want to take a look :)

chetnieter commented on arrows/ffmpeg/ffmpeg_video_input.cxx:
```diff
@@ -0,0 +1,749 @@
+ffmpeg_video_input
+::end_of_video() const
+{
+  return false;
```
Is there a way to tell from the FFmpeg API if we are at the end of the video? It looks like the old VXL implementation set a flag to mark it at the end if the decoding failed but I am not sure if that is the correct approach.

chetnieter replied:
Please take a look at the VXL implementation and reproduce here since the `end_of_video()` method is used in multiple places in both maptk and kwiver.

chetnieter commented on arrows/ffmpeg/tests/test_open_video.cxx:
```diff
@@ -0,0 +1,167 @@
+/*ckwg +29
```
Maybe rename file to `test_video_input_ffmpeg.cxx` since the tests cover more than just opening the video.

chetnieter commented on arrows/ffmpeg/tests/test_video_input_ffmpeg.cxx:
```diff
@@ -0,0 +1,209 @@
+    EXPECT_FALSE(input.seek_frame(ts, requested_frame));
+    EXPECT_NE(requested_frame, ts.get_frame());
+  }
+}
```
Please reproduce the "is_good" and "read_video" tests from the VXL video reader [tests](https://github.com/Kitware/kwiver/blob/master/arrows/vxl/tests/test_vidl_ffmpeg_video_input.cxx).

vovythevov commented:
@chetnieter: Bump :)

chetnieter commented on arrows/ffmpeg/ffmpeg_video_input.cxx:
```diff
@@ -0,0 +1,826 @@
+  *
+  * @return \b true if video was valid and we found a frame.
+  */
+  virtual bool advance()
```
Is there a reason this method in the private data class is virtual? It is causing compilation to fail with clang on MacOS.

linus-sherrill replied:
If it is the private class it seems unlikely that it will be used as a base class. I would agree with removing the virtual.

chetnieter commented on arrows/ffmpeg/ffmpeg_video_input.cxx:
```diff
@@ -0,0 +1,749 @@
+        pixel_trait = vital::image_pixel_traits_of<bool>();
+        direct_copy = true;
+        break;
+      }
```
The lack of a `default` is producing a warning. "warning: 167 enumeration values not handled in switch: ...". Could we add a default block that just sets `direct_copy = false` to get rid of these compiler warnings.

chetnieter commented on arrows/ffmpeg/ffmpeg_video_input.cxx:
```diff
@@ -0,0 +1,826 @@
+      return false;
+    }
+
+    // \todo - num_frames not implemented
```
I think this should just be removed since `num_frames` is not implemented. It probably makes sense to add a TODO or better yet open an issue to remind us to improved how the number of frames is reported but this comment is confusing since `num_frames` is not available.

chetnieter commented on arrows/ffmpeg/ffmpeg_video_input.cxx:
```diff
@@ -0,0 +1,826 @@
+    av_freep(&d->f_video_encoding->opaque);
+  }
+
+  //d->num_frames_ = -2;
```
Remove this is it is no longer relevant nor is it like to be used in the future.

chetnieter commented:
Found a few new minor issues but other than that things look good. I noticed there are some outstanding issues raised by others, notable @linus-sherrill and @dstoup that haven't been resolved. Given the funding situation I think it is ok we push dealing with those after merging but I think we should either open a issue or mark with TODO's so they don't get forgotten.

vovythevov commented:
@chetnieter: Thanks for the feedback !
I updated my code to reflect your comments.
As for @linus-sherrill remark's, I rebased the code which and added all the new necessary methods. Is that what I needed to do ?
As for @dstoup remark's, I did not want to change the FindFFMPEG as it doesn't seem that other Find* prime their paths with Fletch variables.

vovythevov commented:
@mleotta: I pushed the algorithm capabilities commit.

mleotta commented on arrows/ffmpeg/ffmpeg_video_input.cxx:
```diff
@@ -0,0 +1,832 @@
+    if (direct_copy)
+    {
+      int size = avpicture_get_size(enc->pix_fmt, width, height);
+      d->current_image_memory = vital::image_memory_sptr(new vital::image_memory(size));
```
it would be safer to use ` std::make_shared<vital::image_memory>(size);` here.

Longer term we might want to explore recycling image memory instead of allocating it for each frame.  However, that is a bit tricky and should certainly wait until later.

vovythevov replied:
Solved with https://github.com/Kitware/kwiver/pull/533/commits/6983d5a90f5ebbe5c0bd184b73c33d002f35482d

mleotta commented on arrows/ffmpeg/ffmpeg_video_input.cxx:
```diff
@@ -0,0 +1,832 @@
+    else
+    {
+      int size = width * height * num_pixels;
+      d->current_image_memory = vital::image_memory_sptr(new vital::image_memory(size));
```
it would be safer to use ` std::make_shared<vital::image_memory>(size);` here.

mleotta commented on arrows/ffmpeg/ffmpeg_video_input.cxx:
```diff
@@ -0,0 +1,832 @@
+  {
+    int width = enc->width;
+    int height = enc->height;
+    int num_pixels = 3;
```
`num_pixels` is a bit misleading for this variable name.  I would call it `num_channels` or `depth`.

vovythevov replied:
Good catch ! Solved with https://github.com/Kitware/kwiver/pull/533/commits/a1384a87ec522d79109bf6abf29f1515c2ff40a1

mleotta commented on arrows/ffmpeg/ffmpeg_video_input.cxx:
```diff
@@ -0,0 +1,832 @@
+      case AV_PIX_FMT_MONOBLACK:
+      {
+        num_pixels = 1;
+        pixel_trait = vital::image_pixel_traits_of<bool>();
```
I'm not sure if mono images are actually stored as an array of `bool` or or not.  It would be interesting to test.  I don't know that we have any test data though.

mleotta commented on arrows/ffmpeg/ffmpeg_video_input.h:
```diff
@@ -0,0 +1,106 @@
+
+  virtual bool seekable() const;
+  virtual size_t num_frames() const
+    {
```
remove extra indentation here.

vovythevov replied:
Solved with https://github.com/Kitware/kwiver/pull/533/commits/a1384a87ec522d79109bf6abf29f1515c2ff40a1

mleotta commented on arrows/ffmpeg/ffmpeg_video_input.cxx:
```diff
@@ -0,0 +1,832 @@
+  // We don't always have all components of a timestamp, so start with
+  // an invalid TS and add the data we have.
+  kwiver::vital::timestamp ts;
+  ts.set_frame(d->frame_number() + d->f_frame_number_offset + 1);
```
Do not use this `frame_number()` function to compute the current frame.   It is not reliable.  We should keep an internal counter indicating the last frame we stepped to and return that instead.

vovythevov replied:
@mleotta: Do you want to solve this before landing the branch ?

mleotta replied:
Let's land the branch as is and the follow up with improvements

mleotta commented on arrows/ffmpeg/CMakeLists.txt:
```diff
@@ -0,0 +1,49 @@
+include_directories(${FFMPEG_INCLUDE_DIR})
+
+if (NOT FFMPEG_FOUND_SEVERAL)
+  message( FATAL_ERROR "FFMPEG headers all in the same directories, "
```
This change could be done in a follow-up branch.

vovythevov commented on arrows/ffmpeg/ffmpeg_video_input.cxx:
```diff
@@ -0,0 +1,832 @@
+  bool number_of_frames_valid;
+
+  // ==================================================================
+  /*
```
https://github.com/Kitware/kwiver/pull/533/commits/a1384a87ec522d79109bf6abf29f1515c2ff40a1
